File: 例子/Java的例子/pure/VerifySESeal.java
import java.io.File;
import java.io.FileInputStream;
import java.util.Arrays;
import java.util.Date;

import net.netca.pki.SEStamp;
import net.netca.pki.encoding.Hex;
import net.netca.pki.encoding.asn1.pki.AlgorithmIdentifier;
import net.netca.pki.encoding.asn1.pki.Hashable;
import net.netca.pki.encoding.asn1.pki.JCEHasher;
import net.netca.pki.encoding.asn1.pki.JCEVerifier;
import net.netca.pki.encoding.asn1.pki.X509Certificate;
import net.netca.pki.encoding.asn1.pki.seseal.CertDigestList;
import net.netca.pki.encoding.asn1.pki.seseal.CertDigestObj;
import net.netca.pki.encoding.asn1.pki.seseal.ESPictureInfo;
import net.netca.pki.encoding.asn1.pki.seseal.ESPropertyInfo;
import net.netca.pki.encoding.asn1.pki.seseal.ExtData;
import net.netca.pki.encoding.asn1.pki.seseal.ExtensionDatas;
import net.netca.pki.encoding.asn1.pki.seseal.Header;
import net.netca.pki.encoding.asn1.pki.seseal.Seal;
import net.netca.pki.encoding.asn1.pki.seseal.Stamp;
import net.netca.pki.encoding.asn1.pki.seseal.StampInfo;


public class VerifySESeal {

	public static void main(String[] args) throws Exception{
		if(args.length!=1 && args.length!=2) {
            System.out.println("Usage: java VerifySESeal sealFileName [dataFileName]");
            return;
        }

		String sealFileName=args[0];
		byte[] sealData=readFile(sealFileName);
		Seal seal=Seal.decode(sealData);
		Date tsaTime=null;
		try {
			if(seal.hasTimeStamp()==false){
				System.out.println("no timestamp");
			} else {
				if(seal.verifyTimeStamp()){
					System.out.println("verify timestamp ok");
					tsaTime=seal.getTimeStampTime();
					System.out.println("timestamp time:"+tsaTime);
					System.out.println("timestamp cert:"+seal.getTimeStampCert().getSubject().getLdapName());
				}else {
					System.out.println("verify timestamp fail");
				}
				
			} 
			if(seal.verify(tsaTime,new JCEVerifier(),new JCEHasher())==false) {
				System.out.println("verify fail");
			} else {
				System.out.println("verify ok");
			} 
		} catch(Exception e) {
			System.out.println("verify fail,errMsg:"+e.getMessage());
		}
		byte[] hashValue=seal.getDataHash();
		if(args.length==2) {
			AlgorithmIdentifier hashAlgo=seal.getHashAlgorithm();
			Hashable hasher=new JCEHasher();
			byte[] hashValue2=hashFile(hashAlgo,hasher,args[1]);
			if(Arrays.equals(hashValue, hashValue2)) {
				System.out.println("hash match");
			} else {
				System.out.println("hash mismatch");
			}
		}
		int format=seal.getSealType();
		if(format==Seal.GBT_38540){
			System.out.println("GB/T 38540 seal");
		} else if(format==Seal.GMT_0031){
			System.out.println("GM/T 0031 seal");
		} else {
			System.out.println("bad seal format");
		}
		System.out.println("version:"+seal.getVersion());
		Stamp stamp=seal.getStamp();
		format=stamp.getStampType();
		if(format==Stamp.GBT_38540){
			System.out.println("GB/T 38540 stamp");
		} else if(format==Stamp.GMT_0031){
			System.out.println("GM/T 0031 stamp");
		} else {
			System.out.println("bad stamp format");
		}
		StampInfo info=stamp.getEstampInfo();
		Header header=info.getHeader();
		System.out.println("stamp version:"+header.getVersion());
		System.out.println("stamp vid:"+header.getVid());
		System.out.println("stamp id:"+info.getEsID());
		
		ESPropertyInfo property=info.getProperty();
		System.out.println("stamp type:"+property.getType());
		System.out.println("stamp name:"+property.getName());
		int certListType=property.getCertListType();
		if(certListType==SEStamp.CERTLIST_TYPE_CERT){
			X509Certificate[] certs=property.getCertList();
			System.out.println("seal signer cert count:"+certs.length);
			for(int i=0;i<certs.length;i++) {
				System.out.println("\tstamp cert "+i+":"+certs[i].getSubject().getLdapName());
			}
		} else {
			CertDigestList certDigestList=property.getCertDigestList();
			CertDigestObj certDigest;
			int count=certDigestList.size();
			System.out.println("seal signer cert digest count:"+certDigestList.size());
			for(int i=0;i<count;i++){
				certDigest=certDigestList.get(i);
				System.out.println("\tstamp cert digest "+i+" type:"+certDigest.getType()+",stamp cert digest "+i+" value:"+Hex.encode(true,certDigest.getValue()));
			}
		}
		System.out.println("stamp createData:"+property.getCreateDate());
		System.out.println("stamp validStart:"+property.getValidStart());
		System.out.println("stamp validEnd:"+property.getValidEnd());
		
		ESPictureInfo picture=info.getPicture();
		System.out.println("stamp picture type:"+picture.getType());
		System.out.println("stamp picture width:"+picture.getWidth());
		System.out.println("stamp picture height:"+picture.getHeight());
				
		ExtensionDatas exts=info.getExtDatas();
		if(exts==null) {
			System.out.println("stamp no extension");
		} else {
			int count=exts.size();
			ExtData ext;
			System.out.println("stamp extension count:"+count);
			for(int i=0;i<count;i++) {
				System.out.println("stamp extension "+i);
				ext=exts.get(i);
				System.out.println("\textnID:"+ext.getOid());
				System.out.println("\tcritical:"+ext.isCritical());
				System.out.println("\textnValue:"+Hex.encode(true, ext.getExtensionValue()));
			}
		}
		
		System.out.println("stamp maker cert:"+stamp.getCert().getSubject().getLdapName());
		System.out.println("stamp signAlgorithm:"+stamp.getSignatureAlgorithm());
		
		Date time=seal.getTime();
		System.out.println("time:"+time);
		System.out.println("dataHash:"+Hex.encode(true, hashValue));
		System.out.println("propertyInfo:"+seal.getPropertyInfo());
		
		exts=seal.getExtensions();
		if(exts==null) {
			System.out.println("seal no extension");
		} else {
			int count=exts.size();
			ExtData ext;
			System.out.println("seal extension count:"+count);
			for(int i=0;i<count;i++) {
				System.out.println("seal extension "+i);
				ext=exts.get(i);
				System.out.println("\textnID:"+ext.getOid());
				System.out.println("\tcritical:"+ext.isCritical());
				System.out.println("\textnValue:"+Hex.encode(true, ext.getExtensionValue()));
			}
		}
		
		System.out.println("sign cert:"+seal.getCert().getSubject().getLdapName());
		System.out.println("signAlgorithm:"+seal.getSignatureAlgorithm());
		
		
	}

	private static byte[] hashFile(AlgorithmIdentifier hashAlgo,Hashable hasher,String fileName) throws Exception{
		FileInputStream fIn=new FileInputStream(fileName);
		byte[] data=new byte[8192];
		int length;
		hasher.init(hashAlgo);
		try {
			for(;;) {
				length=fIn.read(data);
				if(length==-1) {
					break;
				}
				hasher.update(data, 0, length);
			}
			return hasher.digest();
		}finally {
			fIn.close();
		}
	}
	private static byte[] readFile(String fileName) throws Exception{
		File file=new File(fileName);
		
		long fileLen=file.length();
		if(fileLen>=java.lang.Integer.MAX_VALUE) {
			throw new Exception("file too big:"+ fileName);
		}
		int len=(int)fileLen;
		byte[] data=new byte[len];
		FileInputStream fIn=new FileInputStream(file);
		
		try{
			if (fIn.read(data)!=len) {
				throw new Exception("read file fail:" + fileName);
			}
			return data;
		} finally {
			fIn.close();
		}
	}
	
}

File: 例子/lua的例子/VerifySESeal.lua
local CryptCore=require("CryptCore")

function ToANSIString(utf8String)
    return CryptCore.UTF8ToACP(utf8String)
end


function Usage(filename)
    print(string.format("Usage: lua %s tbsFileName sealFileName",filename))
    print(ToANSIString("参数"))
    print(ToANSIString("\ttbsFileName：原文的文件名"))
    print(ToANSIString("\tsealFileName：电子签章的文件名"))
    print(ToANSIString("说明"))
    print(ToANSIString("\t验证电子签章"))
end

function GetFileConetent(fileName)
    local fIn=io.open(fileName,"rb")
    if fIn==nil then
        return nil
    end
    local data=fIn:read("*a")
    fIn:close()
    
    return data
end

function HashFile(hashObj,fileObj)
    local data
	local BLOCK_LEN=8192
	while (true) do
 		data=fileObj:read(BLOCK_LEN)
		if not data then
			break
		end

		hashObj:Update(data)
	end
    return hashObj:Final()
end


function GetHashLength(hashAlgo)
	if hashAlgo==0x1000 then
		return 16
	elseif hashAlgo==0x2000 then
		return 20
	elseif hashAlgo==0x3000 then
		return 28
	elseif hashAlgo==0x4000 then
		return 32
	elseif hashAlgo==0x5000 then
		return 48
	elseif hashAlgo==0x6000 then
		return 64
	elseif hashAlgo==0x7000 then
		return 32
	elseif hashAlgo==0x8000 then
		return 28
	elseif hashAlgo==0x9000 then
		return 32
	elseif hashAlgo==0xA000 then
		return 28
	elseif hashAlgo==0xB000 then
		return 32
	elseif hashAlgo==0xC000 then
		return 48
	elseif hashAlgo==0xD000 then
		return 64
	else
		return -1
	end
end

if #arg~=2 then
    Usage(arg[0])
    os.exit(-1)
end

local tbsFileName=arg[1]
local sealFileName=arg[2]
local sealEncode=GetFileConetent(sealFileName)
if sealEncode==nil then
    print(ToANSIString("读电子签章文件失败"));
    os.exit(-1)
end

local seal=CryptCore.NewSeal(sealEncode)
if seal==nil then
    print(ToANSIString("不是电子签章文件"));
    os.exit(-1)
end

local format=seal:GetFormat()
if format==CryptCore.Const["GB/T 38540"] then
    print(ToANSIString("符合GB/T 38540的电子签章"))
else
    print(ToANSIString("符合GM/T 0031的电子签章"))
end

print(ToANSIString("版本号：".. seal:GetVersion()))

local stamp=seal:GetStamp()
print(ToANSIString("电子印章版本号：".. stamp:GetVersion()))
print(ToANSIString("电子印章Vid：".. stamp:GetVid()))
print(ToANSIString("电子印章Id：".. stamp:GetId()))
print(ToANSIString("电子印章类型：".. stamp:GetType()))
print(ToANSIString("电子印章名称：".. stamp:GetName()))
local certListType=stamp:GetCertListType()
if certListType==CryptCore.Const.STAMP_CERTLIST_TYPE_CERT then
    local certs=stamp:GetCertList()
    for i,cert in ipairs(certs) do
        print(ToANSIString("电子印章证书"..i..":" ..cert:GetSubject()))
        cert:Free()
    end
else
    local certDigestList=stamp:GetCertDigestList()
    for i,certDigest in ipairs(certDigestList) do
        print(ToANSIString("电子印章证书"..i.."的杂凑类型:"..certDigest.type..",杂凑值:" ..CryptCore.HexEncode(certDigest.value)))
    end
end

print(ToANSIString("电子印章创建时间：".. stamp:GetCreateDate()))
print(ToANSIString("电子印章有效期开始时间：".. stamp:GetValidStart()))
print(ToANSIString("电子印章有效期结束时间：".. stamp:GetValidEnd()))

local pictureType,pictureData,pictureWidth,pictureHeigth=stamp:GetPicture()
print(ToANSIString("电子印章图片类型：".. pictureType))
print(ToANSIString("电子印章图片宽度：".. pictureWidth))
print(ToANSIString("电子印章图片高度：".. pictureHeigth))


local exts=stamp:GetExtension()
if #exts~=0 then
    local oid
    local critical
    local extValue
    for i,ext in ipairs(exts) do
        oid=ext[1]
        critical=ext[2]
        extValue=ext[3]
        if critical then
            print(ToANSIString(string.format("电子印章关键扩展(%s)：%s", oid,CryptCore.HexEncode(extValue))))
        else
            print(ToANSIString(string.format("电子印章非关键扩展(%s)：%s", oid,CryptCore.HexEncode(extValue))))
        end
    end
else
    print(ToANSIString("电子印章没有扩展"))
end

local signCert=stamp:GetSignCert()
print(ToANSIString("电子印章制章人证书：".. signCert:GetSubject()))
signCert:Free()

print(ToANSIString("电子印章签名算法：".. stamp:GetSignAlgorithm()))
stamp:Free()

print(ToANSIString("签名时间：".. seal:GetTime()))

local hashValue=seal:GetDataHash()
local signAlgo=seal:GetSignAlgorithm()
local hashAlgo=CryptCore.GetHashAlgorithmFromSignAlgorithm(signAlgo)

if tbsFileName~="" then
	local fIn=io.open(tbsFileName,"rb")
	if fIn==nil then
		seal:Free()
		print(ToANSIString("打开原文文件失败"));
		os.exit(-1)
	end

	local hashObj=assert(CryptCore.NewHash(hashAlgo))
	local hashValue2=HashFile(hashObj,fIn)
	fIn:close()
	hashObj:Free()

	if hashValue~=hashValue2 then
		seal:Free()
		print(ToANSIString("原文的Hash验证失败"));
		os.exit(-1)
	end
else
	if GetHashLength(hashAlgo)~=#hashValue then
		seal:Free()
		print(ToANSIString("原文的Hash长度验证失败"));
		os.exit(-1)
	end
end
print(ToANSIString("原文属性：".. seal:GetProperty()))

exts=seal:GetExtension()
if #exts~=0 then
    local oid
    local critical
    local extValue
    for i,ext in ipairs(exts) do
        oid=ext[1]
        critical=ext[2]
        extValue=ext[3]
        if critical then
            print(ToANSIString(string.format("电子签章关键扩展(%s)：%s", oid,CryptCore.HexEncode(extValue))))
        else
            print(ToANSIString(string.format("电子签章非关键扩展(%s)：%s", oid,CryptCore.HexEncode(extValue))))
        end
    end
else
    print(ToANSIString("电子签章没有扩展"))
end

local signCert=seal:GetSignCert()
print(ToANSIString("签名证书：" ..signCert:GetSubject()))
signCert:Free()

print(ToANSIString("签名算法：".. signAlgo))

local tsaTime=nil

local hasTimeStamp=seal:HasTimeStamp()
if hasTimeStamp then
    if seal:VerifyTimeStamp() then
        print(ToANSIString("验证电子签章的时间戳成功"));
        local timeStampObj=assert(seal:GetTimeStamp())
        tsaTime=timeStampObj:GetTime()
        print(ToANSIString(string.format("时间戳时间：%s",tsaTime)))
        cert=assert(timeStampObj:GetTSACert())
        print(ToANSIString(string.format("时间戳证书:%s",cert:GetSubject())))
        cert:Free()
        timeStampObj:Free()
    else
        print(ToANSIString("验证电子签章的时间戳失败"));
    end
else
    print(ToANSIString("电子签章没有时间戳"))
end

if tsaTime then
    if seal:Verify(tsaTime) then
        print(ToANSIString("验证电子签章成功"));
    else
        print(ToANSIString("验证电子签章失败"));
    end
else 
    if seal:Verify() then
        print(ToANSIString("验证电子签章成功"));
    else
        print(ToANSIString("验证电子签章失败"));
    end
end

seal:Free()
